library/core/src/sync/atomic.rs

```diff
@@ -79,6 +79,42 @@
 //!
 //! [lock-free]: https://en.wikipedia.org/wiki/Non-blocking_algorithm
 //!
+//! # Atomic accesses to read-only memory
+//!
+//! In general, *all* atomic accesses on read-only memory are Undefined Behavior. For instance, attempting
+//! to do a `compare_exchange` that will definitely fail (making it conceptually a read-only
+//! operation) can still cause a page fault if the underlying memory page is mapped read-only. Since
+//! atomic `load`s might be implemented using compare-exchange operations, even a `load` can fault
+//! on read-only memory.
+//!
+//! (For the purpose of this section, "read-only memory" is defined as memory that is read-only in
+//! the underlying target, i.e., the pages are mapped with a read-only flag and any attempt to write
+//! will cause a page fault. In particular, an `&u128` reference that points to memory that is
+//! read-write mapped is *not* considered to point to "read-only memory". In Rust, almost all memory
+//! is read-write; the only exceptions are memory created by `const` items or `static` items without
+//! interior mutability, and memory that was specifically marked as read-only by the operating
+//! system via platform-specific APIs.)
+//!
+//! However, as an exception from this general rule, "sufficiently small" atomic loads are
+//! implemented in a way that works on read-only memory. The exact threshold for what makes a load
+//! "sufficiently small" varies depending on the architecture and feature flags, but Rust guarantees
+//! that atomic loads that do not exceed the size documented in the following table are guaranteed
+//! to be read-only:
+//!
+//! | Target architecture | Atomic loads no larger than this are guaranteed read-only |
+//! |---------------|---------|
+//! | `x86`         | 4 bytes |
+//! | `x86_64`      | 8 bytes |
+//! | `arm`         | 4 bytes |
+//! | `aarch64`     | 8 bytes |
+//! | `riscv32`     | 4 bytes |
+//! | `riscv64`     | 8 bytes |
+//! | `powerpc64`   | 8 bytes |
```

Comment:
@programmerjake I have generalized this from powerpc64le to powerpc64 since that seems to be the target_arch name.

Reply:
those are actually different because they use different major revisions of the ABI (as well as being different endian).

Reply:
Yeah I knew about endianess. But do they all have readonly 8-byte atomic loads?

Reply:
yes, but some have 16-byte atomic loads by default too. In particular, `powerpc64le-unknown-linux-gnu` -- LLVM generates a call to `libatomic` which is just the 16-byte atomic load instruction `lq` plus some fences.

Reply:
Yeah the intention of this table is to find reliable lower bounds. We can make better promises for particular situations later, I'd like to start with a reasonable baseline.